From 5e515548f3f882866ed20f849469fdd3a59d283f Mon Sep 17 00:00:00 2001
From: Alexia Death <alexiadeath@gmail.com>
Date: Sat, 26 Dec 2009 18:22:41 +0200
Subject: [PATCH 2/2] upexonly.c: Fix a vertical distortion in image data

In scan result every other column was distorted by 2 pixels down shift.
This patch adds correction to this distortion resulting in a clearer image.
---
 libfprint/drivers/upeksonly.c |   19 +++++++++++++++++++
 1 files changed, 19 insertions(+), 0 deletions(-)

diff --git a/libfprint/drivers/upeksonly.c b/libfprint/drivers/upeksonly.c
index 09763f7..bc8392f 100644
--- a/libfprint/drivers/upeksonly.c
+++ b/libfprint/drivers/upeksonly.c
@@ -159,6 +159,7 @@ static void handoff_img(struct fp_img_dev *dev)
 	size_t size = IMG_WIDTH * sdev->num_rows;
 	struct fp_img *img = fpi_img_new(size);
 	GSList *elem = sdev->rows;
+
 	size_t offset = 0;
 
 	if (!elem) {
@@ -169,6 +170,24 @@ static void handoff_img(struct fp_img_dev *dev)
 	fp_dbg("%d rows", sdev->num_rows);
 	img->height = sdev->num_rows;
 
+/* Every other column in scan is shifted down by 2 pixels. Correction of this distortion
+ * is slighly expesnsive, and it works tolerably without it,
+ * but the image resulting from the read is vertically distorted.
+ */
+	if(sdev->num_rows >  2) {
+		GSList *elem_dest  = sdev->rows;
+		GSList *elem_src = g_slist_nth(sdev->rows, 1);
+
+		do {
+			for (int offs = 1; offs < IMG_WIDTH; offs=offs + 2) {
+				memcpy(elem_dest->data + offs, elem_src->data + offs, 1);
+			}
+
+			elem_dest= g_slist_next(elem_dest);
+
+		} while ((elem_src = g_slist_next(elem_src)) != NULL);
+	}
+	
 /* The scans from this device are rolled right by two colums
  * It feels a lot smarter to correct here than mess with it at
  * read time*/
-- 
1.6.3.3
